tests/test_provider.py

```diff
@@ -51,3 +51,40 @@ def flaky_api(action, path=None, url=None, data=None, headers=None, json=None):
     new_len = new_fh.tell()
     assert new_len == file_size #nosec
 
+def test_url_encoding(provider):
+    expected_paths = []
```

Comment:
I think the testing isn't complete. This needs to confirm that the file was actually created with the correct name, which it doesn't appear to do. Also, it looks like you haven't tested every special character. It looks like you are covering space, `'` and '#'. Please investigate all the special URL characters and ensure you've included them all. 

This sort of begs the question of my earlier comment: Why didn't our existing special character test fail on these characters already, and should this test be expanded to include all of the cases covered in that test, and then we could pick the better test, and use that instead of both?

Reply:
The `test_special_characters` test does not catch this case because the filenames are formatted as `/fn-*` and therefore tests that the filename `/fn-#`. However, the condition that the commercial team encountered and Ilya and I saw in dev testing consisted of a filename with "#" and a space. This scenario causes problems within the graph api and throws an error if the path is left unencoded.

I think this test is testing something fundamentally different than just special characters. This is a test of _url encoding_ and uses this `#`  + `space` example to try to tease out any case where we aren't url encoding paths because it's a known case where microsoft will throw an error if left unencoded. I think both tests test different things and add value. As for the `'`, microsoft deals with apostrophes in a strange way and I wanted to ensure that we were dealing with this case robustly

Reply:
So microsoft accepts a `#`, and they accept a [space], and they accept `#a[space]` but not `#[space]a`?

Reply:
Anecdotally, [space]#a has been the problematic scenario

Reply:
So if a sequence of two characters can break, I think that is a possibility to improve the other test, by confirming every sequence of two characters (not in this PR obviously :-). Such a test would have at most 96*95 = 9,120 2 char combinations, and if we created a bunch of files with 100 combinations each, that's around 90 files to test them all. A little expensive from a run-time perspective, but doesn't necessarily need to be run on every commit, and very easy to code. If microsoft has other weird issues, or if another provider would benefit from url-encoding, it would be nice to have the coverage to be more confident we've caught that

Reply:
Ya that seems doable. Could be something we do in the cloud nightly tests

Reply:
Reed has an MR up on the main cloud that addresses some of this.